portality/bll/services/article.py

```diff
@@ -0,0 +1,375 @@
+from portality.lib.argvalidate import argvalidate
+from portality import models
+from portality.bll import exceptions
+from portality.ui.messages import Messages
+
+from datetime import datetime
+
+class ArticleService(object):
+
+    def batch_create_articles(self, articles, account, duplicate_check=True, merge_duplicate=True, limit_to_account=True):
+        """
+        Create a batch of articles in a single operation.  Articles are either all created/updated or none of them are
+
+        This method checks for duplicates within the provided set and within the current database (if you set duplicate_check=True)
+
+        :param articles:  The list of article objects
+        :param account:     The account creating the articles
+        :param duplicate_check:     Whether to check for duplicates in the batch and in the index
+        :param merge_duplicate:     Should duplicates be merged.  If set to False, this may raise a DuplicateArticleException
+        :param limit_to_account:    Should the ingest be limited only to articles for journals owned by the account.  If set to True, may result in an IngestException
+        :return: a report on the state of the import: {success: x, fail: x, update: x, new: x, shared: [], unowned: [], unmatched: []}
+        """
+        # first validate the incoming arguments to ensure that we've got the right thing
+        argvalidate("batch_create_article", [
+            {"arg": articles, "instance" : list, "allow_none" : False, "arg_name" : "articles"},
+            {"arg": account, "instance" : models.Account, "allow_none" : False, "arg_name" : "account"},
+            {"arg" : duplicate_check, "instance" : bool, "allow_none" : False, "arg_name" : "duplicate_check"},
+            {"arg" : merge_duplicate, "instance" : bool, "allow_none" : False, "arg_name" : "merge_duplicate"},
+            {"arg" : limit_to_account, "instance" : bool, "allow_none" : False, "arg_name" : "limit_to_account"}
+        ], exceptions.ArgumentException)
+
+        # 1. dedupe the batch
+        if duplicate_check:
+            batch_duplicates = self._batch_contains_duplicates(articles)
+            if batch_duplicates:
+                report = {"success" : 0, "fail" : len(articles), "update" : 0, "new" : 0, "shared" : [], "unowned" : [], "unmatched" : []}
+                raise exceptions.IngestException(message=Messages.EXCEPTION_ARTICLE_BATCH_DUPLICATE, result=report)
+
+        # 2. check legitimate ownership
+        success = 0
+        fail = 0
+        update = 0
+        new = 0
+        all_shared = set()
+        all_unowned = set()
+        all_unmatched = set()
+
+        for article in articles:
+            result = self.create_article(article, account, duplicate_check=duplicate_check, merge_duplicate=merge_duplicate, limit_to_account=limit_to_account, dry_run=True)
+            success += result.get("success", 0)
```

Comment:
since we enforce the structure of `result` below, we shouldn't need to do this careful value extraction. No harm done, though!